Halt waits in h() when key is released - enables different actions depending on key press length

TRPB opened the pull request:
I created  an issue #120 for this but I was able to implement it myself.

This enables the creation of macros which do different things depending on how long the button has been pressed:

# Functionality

## Long press / short press

```
set("pressed", 1).h(w(500).set("pressed",0)).ifeq("pressed", 1, k(x), k(y))
```

If the button is released within 500ms it prints `x` if it's released after 500ms it prints `y`. 


## Tap / hold

```
set("pressed", 1).h(w(100).set("pressed",0).h(y)).ifeq("pressed", 1, k(x), k(y))
```

On tap (button released within 100ms) prints `x` after 100ms it acts like `y` is being held down. 

# Implementation

The main change is in the `run` method:


```python
        for task in self.tasks:
            # one could call tasks the compiled macros. it's lambda functions
            # that receive the handler as an argument, so that they know
            # where to send the event to.
            coroutine = task(handler)

            if (self.is_holding() == False and coroutine.__name__ == "sleep"):
                break

            if asyncio.iscoroutine(coroutine):
                await coroutine
                if (self.is_holding() == False and coroutine.__name__ == "sleep"):
                    break
```

If the button has been released and the next task is `wait`   then it stops running any further tasks. E.g. if the button is released during the wait of `w(500).k(y)`  then `k(y)` will never get executed. This is done twice because the key could be released before the loop runs or during the `await`.

The second change is the sleep function:


```py
        async def sleep(_):
            """ Wait in intervals of 10ms so the wait can be ended early if the key is released """
            for i in range(int(sleeptime/10)):
                await asyncio.sleep(0.01)
                if (self.is_holding() == False):
                    break

```

Instead of `await asyncio.sleep(sleeptime/1000)` it chunks the waits into 10ms intervals. Without this in the example above, tapping and releasing the key would make the `x` appear only after the wait timer had finished (e.g. 500ms). This reduces the input lag to 10ms max.

sezanzeb commented:
Thanks for contributing! I'll review your code shortly

This comment is actually deprecated:

> run the child macro completely to avoid not-releasing any key

It was valid when the `def keycode` function was adding multiple tasks, one for key-down, for wait, and for key-up. Stopping the macro early could have potentially stopped it after the key-down, which would result in a key that is held down forever. It was changed to be a single task at some point.

So I think your change works

TRPB commented:
Playing with this further, the first if statement in `run` can be removed as I added that before I modified the sleep function.

I'm not sure how other people are using macros but this change could potentially break some macros. If you had `k(x).w(500).k(x)` with this change, x would only appear twice if the button was held down.

Perhaps it needs an argument? `w(500, True)` would exit if the key is released whereas `w(500, False)` or `w(500)` would keep running the chain after the key had been released?

sezanzeb commented:
that reminds me of https://github.com/sezanzeb/key-mapper/issues/109

if there was more space for editing macros it would be totally fine to add new functions with longer names, or to use meaningful flags like `hold(..., breakable)`.

Not sure if I would find it intuitive that a `w` function has a special meaning for `h`, since I wouldn't expect that if it was treated like a lambda function in python.

sezanzeb commented:
Any updates? There are quite a lot of ways this could be solved in the future and things to consider, so lets discuss it a bit first before making a decision on it

sezanzeb commented:
this is covered with if_tap now
